File: pkg/models/Tag.go
package models

import (
	"strings"
	"time"
)

type Tag struct {
	UUID                   string             `json:"uuid" gorm:"type:uuid;not_null;primary_key"`
	Name                   string             `json:"name" gorm:"size:100;not_null"`
	Mac                    string             `json:"mac"`
	VerificationDate       time.Time          `json:"verification_date"`
	HigherTemperatureLimit float64            `json:"higher_temperature_limit"`
	LowerTemperatureLimit  float64            `json:"lower_temperature_limit"`
	TemperatureZones       []*TemperatureZone `json:"temperature_zones" gorm:"many2many:temperature_zone_tags;"`
	SlaveId                int                `gorm:"-"`
}

func (src Tag) Equals(dst Tag) bool {
	if strings.Compare(src.UUID, dst.UUID) == 0 &&
		strings.Compare(src.Mac, dst.Mac) == 0 &&
		strings.Compare(src.Name, dst.Name) == 0 {
		return true
	} else {
		return false
	}
}

func (Tag) TableName() string {
	return "tag"
}

type TagList []Tag

func (src TagList) Difference(dst TagList, compare func(a Tag, b Tag) bool) TagList {
	var diff TagList
	i := 0
	j := 0
	n := len(src)
	m := len(dst)
	for i < n && j < m {
		if !compare(src[i], dst[j]) {
			diff = append(diff, src[i])
		} else {
			j++
		}
		i++
	}
	for i < n {
		diff = append(diff, src[i])
	}
	return diff
}

func (src TagList) Intersect(dst TagList, compare func(a Tag, b Tag) bool) TagList {
	var intersect TagList
	i := 0
	j := 0
	n := len(src)
	m := len(dst)
	for i < n && j < m {
		if compare(src[i], dst[j]) {
			intersect = append(intersect, src[i])
		} else {
			j++
		}
		i++
	}
	for i < n {
		intersect = append(intersect, src[i])
	}
	return intersect
}

File: fetch_service/src/main/java/ru/tagmeasurements/fetch_service/models/Tag.java
package ru.tagmeasurements.fetch_service.models;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.TypeDef;
import org.hibernate.annotations.TypeDefs;
import ru.tagmeasurements.fetch_service.dtos.TagResponseData;
import ru.tagmeasurements.fetch_service.dtos.TempSensorConfig;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Data
@Entity
@NoArgsConstructor
@AllArgsConstructor
public class Tag {
    @Id
    private UUID uuid;
    private String name;
    private String mac;
    private LocalDate verificationDate;
    private Double higherTemperatureLimit;
    private Double lowerTemperatureLimit;
    @ManyToMany
    private List<TemperatureZone> temperatureZones;

    @Transient
    private Long slaveId;
    @Transient
    private Integer tagType;

    public void parse(TagResponseData response) {
        uuid = UUID.fromString(response.getUuid());
        name = response.getName();
        slaveId = response.getSlaveId();
        tagType = response.getTagType();
    }

    public void parse(TempSensorConfig config) {
        lowerTemperatureLimit = config.getTh_low();
        higherTemperatureLimit = config.getTh_high();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Tag tag = (Tag) o;
        return uuid.equals(tag.uuid) &&
                name.equals(tag.name) &&
                mac.equals(tag.mac) &&
                Objects.equals(verificationDate, tag.verificationDate) &&
                Objects.equals(higherTemperatureLimit, tag.higherTemperatureLimit) &&
                Objects.equals(lowerTemperatureLimit, tag.lowerTemperatureLimit);
    }

    @Override
    public int hashCode() {
        return Objects.hash(uuid, name, mac, verificationDate, higherTemperatureLimit, lowerTemperatureLimit);
    }

}
